libbeat/api/routes.go

```diff
@@ -73,3 +97,57 @@ func prettyPrint(w http.ResponseWriter, data common.MapStr, u *url.URL) {
 		fmt.Fprintf(w, data.String())
 	}
 }
+
+type Routes struct {
+	routes map[string]handlerFunc
+	log    *logp.Logger
+	smux   *http.ServeMux
+	mux    sync.RWMutex
+}
+
+func (d *Routes) handle(w http.ResponseWriter, r *http.Request) {
+	d.mux.RLock()
+	defer d.mux.RUnlock()
+
+	if h, exist := d.routes[r.URL.Path]; exist {
+		h(w, r)
+		return
+	}
+
+	http.NotFound(w, r)
+}
+
+func (d *Routes) register(api string, h handlerFunc) error {
+	d.mux.Lock()
+	defer d.mux.Unlock()
+	if !strings.HasPrefix(api, "/") {
+		return fmt.Errorf("route should starts with /")
```

Comment:
I think it is quite unfortunate that we need to use absolute addresses when registering. This seems to be a byproduct of having Routes global.

At some point we need to think about some proper API for sub-systems to register/deregister routes. Kind of similar to gorilla/mux `SubRoutes`.